### PR TITLE
Add info about ca cert used to verify chain.

### PR DESCRIPTION
**Pull request by gtcooke94**
The tsi_peer object will now contain the subject of the root/ca cert that was used to verify the peer's chain during a handshake.

We have to get this information into `tsi_peer` in a roundabout way - unfortunately BoringSSL does not have the function `SSL_get0_verified_chain`. The only way to get the verified chain (and thus the CA/root cert that was used to verify the peer's chain) is through the `X509_STORE_CTX` used during the handshake post-verification. So we tie into that callback, then we store the CA Cert on the SSL object as `ex_data`. The SSL object is what we have access to when setting the values on `tsi_peer`, thus we need the information on that object.

There was issue with overlap with a CRL-related callback, so I had to move that function further up in the file.

There will be a follow-up PR that adds this information to `grpc_tls_custom_verification_check_request` and `TlsCustomVerificationCheckRequest` (these structs/classes are constructed using information in `tsi_peer`), with the end goal of the grpc-user having access to the Subject information of the CA cert used to verify the peer chain during custom verification callbacks.

The main changes are in the `ssl_transport_security.cc` file.
I added a test that uses a certificate chain with an intermediate CA, so there are several file additions to the `test_creds` directory. The README changes explain how these files are generated.

### CODE REVIEW

**Comment by gtcooke94** [comment]
@matthewstevenson88

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:869]: Should we be returning `ok` instead of `1` here?

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:905]: Shouldn't all return values continue to be `preverify_ok`? We don't want to change the existing behavior, just opportunistically extract this info.

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:909]: Please return early if `ca_cert` is nullptr.

Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:909]: Shouldn't we check that `sk_X509_num` is `>0` before this line?

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:913]: Please check that `ssl` is non-null before this line.

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:1219]: nit: typo in "result". Also can we add a more informative error message here?

Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:1219]: I'm not sure this should be an error log, since this will occur routinely e.g. if a self-signed cert is used. Could we make it a debug log instead?

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:1220]: We shouldn't fail if there is a problem here, I think we should just log and continue.

File: test/core/tsi/ssl_transport_security_test.cc
Comment (matthewstevenson88) [inline test/core/tsi/ssl_transport_security_test.cc:199]: nit: Please use `std::string` here instead of `const char*.

File: test/core/tsi/ssl_transport_security_test.cc
Comment (matthewstevenson88) [inline test/core/tsi/ssl_transport_security_test.cc:375]: Can we check that this peer property is *not set* for resumed sessions?

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:147]: Here and throughout: we seem to always refer to "ca cert", but really we're interested in the root CA cert that was used to verify the chain. Can we rename instead say (e.g.) "verified root cert"? For example, this variable would become `g_ssl_ex_verified_root_cert_index`?

File: test/core/tsi/ssl_transport_security_test.cc
Comment (matthewstevenson88) [inline test/core/tsi/ssl_transport_security_test.cc:194]: IIUC, we are only testing in the case when verification is successful and there are no intermediate CAs in the chain. I think it would be good to test in the following cases as well:

1. When verification fails (e.g. we can use the self-signed cert in the /src/core/tsi/test_creds directory).
2. When verification succeeds but there is >=1 intermediate CA in the chain. (If there are no examples of this in /src/core/tsi/test_creds let me know and I can help you generate them.)

**Review by matthewstevenson88 — COMMENTED** [review]
Thanks for the PR! Left a handful of comments, PTAL and feel free to reach out if you'd like to talk about any of them. :)

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:869]: For completeness, I didn't actually add this, just moved it further up in the file.
But to answer the question, no, these verification callbacks just return 1 on success and 0 on failure

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:905]: Above this is a line for `if preverify_ok == 0 return 0;`, so behaviorally here `preverify_ok` will always be 1, but it could be cleaner to just re-use `preverify_ok`.
One of the intents here was that, if `preverify_ok` is _not_ ok (`== 0`) then there won't be a verified chain so we should not continue.

Thoughts?

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:909]: Good idea, I'll add the additional checks for this and for `nullptr`. I think if `chain` were empty but not `nullptr` this would have an issue

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:1220]: Ack, will change

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:909]: Ack, will change

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:913]: Ack, will add check

File: test/core/tsi/ssl_transport_security_test.cc
Comment (gtcooke94) [reply test/core/tsi/ssl_transport_security_test.cc:199]: `const char* expected match` is used through this test file, [example](https://github.com/grpc/grpc/blob/a4e3a22f5179fabc1670dcfa6d58e4adfaf7b02a/test/core/tsi/ssl_transport_security_test.cc#L322)

File: test/core/tsi/ssl_transport_security_test.cc
Comment (gtcooke94) [reply test/core/tsi/ssl_transport_security_test.cc:375]: So you mean in an `else` block after this, explicitly check that there is no peer property with `TSI_X509_CA_SUBJECT_PEER_PROPERTY`?

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:1219]: We had placed this log here for debugging - do you think we actually need it? Above a lot of the properties follow the pattern `if result != TSI_OK) return result;`. I think the `peer_chain` is the only property above that doesn't return result.

I guess if we reach here, what we're saying is "there is a root cert that is set in the ex_data, but adding it as a property to `tsi_peer` is failing." So what's our desired behavior here?

File: src/core/tsi/ssl_transport_security.cc
Comment (rockspore) [inline src/core/tsi/ssl_transport_security.cc:2217]: I cann't comment right there but should l2207 also have this callback?

Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:2217]: Defer to @matthewstevenson88 on that, it's the case of `TSI_REQUEST_CLIENT_CERTIFICATE_AND_VERIFY`

File: test/core/tsi/ssl_transport_security_test.cc
Comment (gtcooke94) [reply test/core/tsi/ssl_transport_security_test.cc]: Added nullptr checks in the test for cases where this should be unset

Is case 2 a special case? The implementation just gets the last cert in the verified chain, should it matter if there are intermediate CAs?

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc]: Resolving because I think we will address this in this thread: https://github.com/grpc/grpc/pull/32215#discussion_r1089523074

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:488]: nit: Please add an inline comment here /*is_verified_root_cert=*/ before the "false"

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [reply src/core/tsi/ssl_transport_security.cc:869]: Ah I see. Why did it need to be moved?

If there's a good reason to move it, can we leave it where it is in this PR and move it in a follow-up PR?

Rationale: This CL has some subtleties to it, and we want to avoid making unrelated changes that will obscure the real functional changes.

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:888]: Why does this take precendence?

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:896]: Please add a comment explaining the rationale behind this if statement.

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:885]: Please check whether `ctx` is null when we enter the function. If we call `X509_STORE_CTX_get0_chain` and `ctx` is null, then we will segfault.

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [inline src/core/tsi/ssl_transport_security.cc:909]: nit: Please avoid the use of `auto`.

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [reply src/core/tsi/ssl_transport_security.cc:2217]: Yup good call, we should have it in both cases.

File: test/core/tsi/ssl_transport_security_test.cc
Comment (matthewstevenson88) [reply test/core/tsi/ssl_transport_security_test.cc]: Maybe I misunderstood, but I didn't see any tests where verification fails?

I think 2 is important because the intermediates are sent by the peer but the root is not, it's from our trust store.

File: src/core/tsi/ssl_transport_security.cc
Comment (matthewstevenson88) [reply src/core/tsi/ssl_transport_security.cc]: Sounds reasonable.

File: test/core/tsi/ssl_transport_security_test.cc
Comment (gtcooke94) [reply test/core/tsi/ssl_transport_security_test.cc]: There are failed verifications in the [CRL related SSL tests](https://github.com/grpc/grpc/blob/master/test/core/tsi/crl_ssl_transport_security_test.cc)

I'll happily add more tests though, I'll discuss with you offline about setting up the chain for 2.

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:2217]: Adding

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:888]: It's because of this [code in boringssl](https://boringssl.googlesource.com/boringssl/+/refs/heads/master/ssl/ssl_x509.cc#402).

In the chain of BoringSSL, if you set a verification callback with `SSL_CTX_set_verify` it ends up as the callback here on `hs->config->verify_callback`. So if that is set, it will override a callback set on `X509_STORE_CTX_set_verify` with whatever is there.

And this is the case here - we've maybe set a CRL-related verification callback with `X509_STORE_CTX_set_verify`, but we've also set the callback to extract the root cert with `SSL_CTX_set_verify`

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:488]: Will do

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:896]: Will do

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:885]: Will do

File: src/core/tsi/ssl_transport_security.cc
Comment (gtcooke94) [reply src/core/tsi/ssl_transport_security.cc:909]: Changed to `size_t`

File: src/core/tsi/test_creds/README
Comment (matthewstevenson88) [inline src/core/tsi/test_creds/README:18]: nit: Should this line be removed?

File: src/core/tsi/test_creds/README
Comment (matthewstevenson88) [inline src/core/tsi/test_creds/README:103]: nit: This comment is a bit circular, could you clarify? Also please mention the other files (e.g. leaf_and_intermediate_chain.pem) that are being created.

File: src/core/tsi/test_creds/intermediate.cnf
Comment (matthewstevenson88) [reply src/core/tsi/test_creds/intermediate.cnf:12]: Friendly ping :)

File: src/core/tsi/test_creds/leaf_signed_by_intermediate.cnf
Comment (matthewstevenson88) [inline src/core/tsi/test_creds/leaf_signed_by_intermediate.cnf:12]: Same here RE blank line to make the linter happy. :)

**Review by matthewstevenson88 — COMMENTED** [review]
A few miscellaneous comments, and then this CL LGTM. If it's OK with you, I might ask Zhan to take a look as well, because this PR got quite technical.

We should find a reviewer from the gRPC team to take a look as well.

File: src/core/tsi/test_creds/leaf_signed_by_intermediate.cnf
Comment (gtcooke94) [reply src/core/tsi/test_creds/leaf_signed_by_intermediate.cnf:12]: Done

File: src/core/tsi/test_creds/README
Comment (gtcooke94) [reply src/core/tsi/test_creds/README]: Done

File: src/core/tsi/test_creds/README
Comment (gtcooke94) [reply src/core/tsi/test_creds/README]: Done

File: src/core/tsi/test_creds/leaf_signed_by_intermediate.cnf
Comment (gtcooke94) [reply src/core/tsi/test_creds/leaf_signed_by_intermediate.cnf]: Done

File: src/core/tsi/test_creds/intermediate.cnf
Comment (gtcooke94) [reply src/core/tsi/test_creds/intermediate.cnf]: Done

**Comment by gtcooke94** [comment]
Feel free to add Zhan as another pair of eyes, and I believe I've addressed your last round of comments.

There's a few tests failing, but I believe it's some infra/CI related issues (I made no material code changes since the last commit since everything passed). I'll keep an eye on it and re-run the tests.

**Comment by matthewstevenson88** [comment]
@zeromath

**Comment by zeromath** [comment]
LGTM on OpenSSL APIs.

**Comment by gtcooke94** [comment]
@markdroth we've got LGTM from the security side, can we get a general language review?

**Review by markdroth — APPROVED** [review]

**Pull request merged**
